File: Easy/70-Climbing-Stairs.java
// Time: O(N)
// Space: O(1)
// An improved version of dp. Uses fibonacci sequence

class Solution {
    public int climbStairs(int n) {
        if (n == 1) {
            return 1;
        }
        
        int first = 1, second = 1;
        int ans = 0;
        
        for (int i = 0; i < n - 1; i++) {
            ans = first + second;
            first = second;
            second = ans;
        }
        
        return ans;
    }
}

File: Easy/70-Climbing-Stairs.py
class Solution(object):
    def climbStairs(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n == 1:
            return 1

        a, b = 1, 2

        for i in range(n - 2, 0, -1):
            tmp = b
            b = tmp + a
            a = tmp

        return b
